File: traditional_comm_py/traditional_comm/runner.py
from __future__ import annotations

import hashlib
import json
from pathlib import Path
import time
import tracemalloc
import uuid

from .codecs import codec_info, decode_payload, encode_payload
from .samples import parse_ppm, parse_tvid
from .task_adapter import TaskAdapter, create_task_adapter
from .task_metrics import quality_metrics


def create_run_id(kind: str) -> str:
    return f"traditional-{kind}-{int(time.time() * 1000)}-{uuid.uuid4().hex[:6]}"


def sha256(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def _write_json(path: Path, value: dict) -> None:
    path.write_text(json.dumps(value, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")


def _validate(kind: str, output: bytes) -> tuple[bool, dict]:
    if kind == "text":
        output.decode("utf-8")
        return True, {"format": "txt", "encoding": "utf-8"}
    if kind == "image":
        info = parse_ppm(output)
        return True, {
            "format": "ppm",
            "width": info["width"],
            "height": info["height"],
        }
    info = parse_tvid(output)
    return True, {
        "format": "tvid-decoded-frames",
        "width": info["width"],
        "height": info["height"],
        "fps": info["fps"],
        "frame_count": info["frame_count"],
    }


def _quality_metrics(kind: str, source: bytes, output: bytes) -> dict:
    return quality_metrics(kind, source, output)


def _codec_config(kind: str, encoded) -> dict:
    config = codec_info(kind)
    config.update(
        {
            "name": encoded.codec,
            "container": encoded.container,
            "encoder": encoded.metadata.get("encoder"),
            "parameters": encoded.metadata.get("parameters", {}),
            "metadata": encoded.metadata,
        }
    )
    return config


def run_one(
    kind: str,
    input_path: Path,
    runs_dir: Path,
    transport,
    task: dict | None = None,
    run_id: str | None = None,
    codec_options: dict | None = None,
    task_adapter: TaskAdapter | None = None,
) -> dict:
    task = task or {}
    task_adapter = task_adapter or create_task_adapter()
    run_id = run_id or create_run_id(kind)
    run_dir = runs_dir / run_id
    run_dir.mkdir(parents=True, exist_ok=True)
    input_data = input_path.read_bytes()
    input_hash = sha256(input_data)
    tracemalloc.start()
    process_start = time.process_time()
    started_at = time.time()

    encode_start = time.perf_counter()
    encoded = encode_payload(kind, input_data, codec_options)
    encode_ms = (time.perf_counter() - encode_start) * 1000

    transport_start = time.perf_counter()
    received, transport_stats = transport.send_payload(
        encoded.payload,
        {
            "run_id": run_id,
            "mode": "traditional",
            "sample_id": input_path.name,
            "task_id": task.get("task_id"),
            "media_type": kind,
            "codec": encoded.codec,
            "container": encoded.container,
            "codec_metadata": encoded.metadata,
            "payload_size": len(encoded.payload),
            "expected_total_bytes": len(encoded.payload),
            "sequence_id": 0,
            "channel": task.get("channel", {}),
            "nodes": task.get("nodes", {}),
        },
    )
    transport_ms = (time.perf_counter() - transport_start) * 1000

    received_path = run_dir / f"received_payload.{encoded.container}"
    received_path.write_bytes(received)
    encoded_path = run_dir / f"encoded_payload.{encoded.container}"
    encoded_path.write_bytes(encoded.payload)

    decode_start = time.perf_counter()
    output = decode_payload(kind, received, encoded.metadata)
    decode_ms = (time.perf_counter() - decode_start) * 1000
    valid, validation = _validate(kind, output)
    receiver_decode_valid = transport_stats.get("receiver_decode_valid", True)
    valid = bool(valid and receiver_decode_valid is not False)

    output_extension = {"text": "txt", "image": "jpg", "video": "mp4"}[kind]
    output_path = run_dir / f"output.{output_extension}"
    output_path.write_bytes(received)
    decoded_path = run_dir / {"text": "decoded.txt", "image": "decoded.ppm", "video": "decoded.tvid"}[kind]
    decoded_path.write_bytes(output)

    output_hash = sha256(output)
    content_match = input_hash == output_hash if kind == "text" else None
    input_bytes = len(input_data)
    encoded_bytes = len(encoded.payload)
    reduction = (input_bytes - encoded_bytes) / input_bytes if input_bytes else 0
    total_latency_ms = encode_ms + transport_ms + decode_ms
    current, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()
    process_cpu_ms = (time.process_time() - process_start) * 1000
    quality = _quality_metrics(kind, input_data, output)
    task_result = task_adapter.run(
        kind=kind,
        source_path=input_path,
        decoded_path=decoded_path,
        source_bytes=input_data,
        decoded_bytes=output,
        task=task,
        quality=quality,
    )
    task_ok = bool(task_result.get("success", False))

    config = {
        "run_id": run_id,
        "mode": "traditional",
        "media_type": kind,
        "input_path": str(input_path.resolve()),
        "input_sha256": input_hash,
        "sample": {
            "sample_id": input_path.name,
            "path": str(input_path.resolve()),
            "media_type": kind,
            "original_format": input_path.suffix.lstrip("."),
            "checksum": input_hash,
        },
        "task": task,
        "codec": _codec_config(kind, encoded),
        "channel": task.get("channel", {}),
        "nodes": task.get("nodes", {}),
        "system": task.get("system", {}),
        "transport": transport_stats.get("transport", type(transport).__name__),
        "started_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime(started_at)),
    }
    metrics = {
        "run_id": run_id,
        "mode": "traditional",
        "media_type": kind,
        "codec": encoded.codec,
        "container": encoded.container,
        "input_bytes": input_bytes,
        "encoded_payload_bytes": encoded_bytes,
        "actual_sent_bytes": transport_stats["sent_bytes"],
        "actual_received_bytes": transport_stats["received_bytes"],
        "actual_link_total_bytes": transport_stats.get(
            "actual_link_total_bytes",
            transport_stats["sent_bytes"],
        ),
        "data_reduction_ratio": round(reduction, 6),
        "encode_time_ms": round(encode_ms, 3),
        "transport_time_ms": round(transport_ms, 3),
        "decode_time_ms": round(decode_ms, 3),
        "end_to_end_latency_ms": round(total_latency_ms, 3),
        "average_throughput_mbps": transport_stats["average_throughput_mbps"],
        "peak_throughput_mbps": transport_stats["peak_throughput_mbps"],
        "receiver_decode_time_ms": transport_stats.get("receiver_decode_time_ms"),
        "receiver_decode_valid": receiver_decode_valid,
        "receiver_result_path": transport_stats.get("receiver_result_path"),
        "quality": quality,
        "task": {
            "content_match": content_match,
            "output_valid": int(valid),
            "task_result": task_result,
        },
        "resource": {
            "python_process_cpu_ms": round(process_cpu_ms, 3),
            "tracemalloc_current_bytes": current,
            "tracemalloc_peak_bytes": peak,
            "gpu_usage": None,
        },
        "status": "completed" if valid and task_ok else "failed",
    }
    _write_json(run_dir / "config.json", config)
    _write_json(run_dir / "metrics.json", metrics)
    _write_json(run_dir / "transport.json", transport_stats)
    _write_json(
        run_dir / "result.json",
        {
            "run_id": run_id,
            "encoded_path": str(encoded_path),
            "received_path": str(received_path),
            "output_path": str(output_path),
            "decoded_path": str(decoded_path),
            "output_sha256": output_hash,
            "content_match": content_match,
            "validation": validation,
            "receiver_decode_valid": receiver_decode_valid,
            "receiver_result_path": transport_stats.get("receiver_result_path"),
            "task_result": task_result,
        },
    )
    return {
        "run_id": run_id,
        "run_dir": run_dir,
        "kind": kind,
        "metrics": metrics,
        "output_path": output_path,
    }


def run_task_only(
    kind: str,
    input_path: Path,
    runs_dir: Path,
    task: dict | None = None,
    run_id: str | None = None,
    task_adapter: TaskAdapter | None = None,
) -> dict:
    """Run a downstream task directly on a task-compatible input file.

    This is used for feature-level datasets such as MOSEI.  A MOSEI ``.pkl``
    feature sample is not an encoded video container, so it must not be sent
    through the H.264/MP4 transport path.  The result still uses the same
    run directory and task-result schema as a normal communication run.
    """

    task = task or {}
    task_adapter = task_adapter or create_task_adapter()
    run_id = run_id or create_run_id(f"task-{kind}")
    run_dir = runs_dir / run_id
    run_dir.mkdir(parents=True, exist_ok=True)
    input_path = input_path.resolve()
    input_data = input_path.read_bytes()
    task_result = task_adapter.run(
        kind=kind,
        source_path=input_path,
        decoded_path=input_path,
        source_bytes=input_data,
        decoded_bytes=input_data,
        task=task,
        quality={},
    )
    status = "completed" if task_result.get("success") else "failed"
    config = {
        "run_id": run_id,
        "mode": "traditional",
        "execution": "task-only",
        "media_type": kind,
        "input_path": str(input_path),
        "task": task,
        "transport": None,
        "note": "No media codec or transport was run; input is already task-compatible.",
    }
    metrics = {
        "run_id": run_id,
        "mode": "traditional",
        "execution": "task-only",
        "media_type": kind,
        "input_bytes": len(input_data),
        "encoded_payload_bytes": None,
        "actual_sent_bytes": None,
        "actual_received_bytes": None,
        "end_to_end_latency_ms": task_result.get("inference_time_ms"),
        "task": {"task_result": task_result},
        "status": status,
    }
    _write_json(run_dir / "config.json", config)
    _write_json(run_dir / "metrics.json", metrics)
    _write_json(run_dir / "result.json", {"run_id": run_id, "input_path": str(input_path), "task_result": task_result})
    return {"run_id": run_id, "run_dir": run_dir, "kind": kind, "metrics": metrics, "output_path": input_path}


def run_all(
    samples_dir: Path,
    runs_dir: Path,
    transport,
    task: dict | None = None,
    codec_options: dict | None = None,
    task_adapter: TaskAdapter | None = None,
) -> list[dict]:
    runs_dir.mkdir(parents=True, exist_ok=True)
    inputs = {
        "text": samples_dir / "sample.txt",
        "image": samples_dir / "sample.ppm",
        "video": samples_dir / "sample.tvid",
    }
    missing = [str(path) for path in inputs.values() if not path.exists()]
    if missing:
        raise FileNotFoundError("Missing samples: " + ", ".join(missing))
    return [
        run_one(
            kind,
            path,
            runs_dir,
            transport,
            task,
            codec_options=codec_options,
            task_adapter=task_adapter,
        )
        for kind, path in inputs.items()
    ]

File: traditional_comm/src/runner.js
const crypto = require('node:crypto');
const fs = require('node:fs');
const path = require('node:path');

const { codecInfo, decodePayload, encodePayload } = require('./codecs');
const { parseTvid } = require('./samples');

function ensureDir(dir) {
  fs.mkdirSync(dir, { recursive: true });
}

function sha256(buffer) {
  return crypto.createHash('sha256').update(buffer).digest('hex');
}

function elapsedMs(start) {
  return Number(process.hrtime.bigint() - start) / 1e6;
}

function writeJson(filePath, value) {
  fs.writeFileSync(filePath, `${JSON.stringify(value, null, 2)}\n`, 'utf8');
}

function getOutputName(kind, runDir) {
  const extension = kind === 'text' ? 'txt' : kind === 'image' ? 'ppm' : 'tvid';
  return path.join(runDir, `output.${extension}`);
}

function createRunId(kind) {
  return `traditional-${kind}-${Date.now()}-${crypto.randomBytes(3).toString('hex')}`;
}

function validateDecoded(kind, buffer) {
  if (kind === 'text') {
    buffer.toString('utf8');
    return { valid: true, details: {} };
  }
  if (kind === 'image') {
    const header = buffer.subarray(0, 2).toString('ascii');
    return { valid: header === 'P6', details: { format: header } };
  }
  const parsed = parseTvid(buffer);
  return { valid: true, details: parsed };
}

async function runOne({ kind, inputPath, runsDir, transport, task = {}, runId: providedRunId }) {
  const runId = providedRunId || createRunId(kind);
  const runDir = path.join(runsDir, runId);
  ensureDir(runDir);

  const input = fs.readFileSync(inputPath);
  const inputHash = sha256(input);
  const startedAt = new Date().toISOString();

  const encodeStart = process.hrtime.bigint();
  const encoded = encodePayload(kind, input);
  const encodeMs = elapsedMs(encodeStart);

  const transportStart = process.hrtime.bigint();
  const transmitted = await transport.sendPayload(encoded.payload, {
    run_id: runId,
    mode: 'traditional',
    sample_id: path.basename(inputPath),
    task_id: task.task_id || null,
    codec: encoded.codec,
    container: encoded.container,
    payload_size: encoded.payload.length,
  });
  const transportMs = elapsedMs(transportStart);

  const decodeStart = process.hrtime.bigint();
  const output = decodePayload(kind, transmitted.payload);
  const decodeMs = elapsedMs(decodeStart);
  const validation = validateDecoded(kind, output);

  const outputPath = getOutputName(kind, runDir);
  fs.writeFileSync(outputPath, output);

  const outputHash = sha256(output);
  const contentMatch = inputHash === outputHash;
  const endToEndMs = encodeMs + transportMs + decodeMs;
  const inputBytes = input.length;
  const encodedBytes = encoded.payload.length;
  const dataReduction = inputBytes === 0 ? 0 : (inputBytes - encodedBytes) / inputBytes;
  const memory = process.memoryUsage();
  const cpu = process.cpuUsage();

  const config = {
    run_id: runId,
    mode: 'traditional',
    media_type: kind,
    input_path: path.resolve(inputPath),
    input_sha256: inputHash,
    task,
    codec: codecInfo(kind),
    transport: 'loopback-or-adapter',
    started_at: startedAt,
    finished_at: new Date().toISOString(),
  };
  const metrics = {
    run_id: runId,
    mode: 'traditional',
    media_type: kind,
    input_bytes: inputBytes,
    encoded_payload_bytes: encodedBytes,
    actual_sent_bytes: transmitted.stats.sent_bytes,
    actual_received_bytes: transmitted.stats.received_bytes,
    data_reduction_ratio: Number(dataReduction.toFixed(6)),
    encode_time_ms: Number(encodeMs.toFixed(3)),
    transport_time_ms: Number(transportMs.toFixed(3)),
    decode_time_ms: Number(decodeMs.toFixed(3)),
    end_to_end_latency_ms: Number(endToEndMs.toFixed(3)),
    average_throughput_mbps: transmitted.stats.average_throughput_mbps,
    peak_throughput_mbps: transmitted.stats.peak_throughput_mbps,
    quality: { psnr: null, ssim: null, lpips: null },
    task: { content_match: contentMatch ? 1 : 0, output_valid: validation.valid ? 1 : 0 },
    resource: {
      rss_bytes: memory.rss,
      heap_used_bytes: memory.heapUsed,
      cpu_user_ms: Math.round(cpu.user / 1000),
      cpu_system_ms: Math.round(cpu.system / 1000),
      gpu_usage: null,
    },
    status: contentMatch && validation.valid ? 'completed' : 'failed',
  };

  writeJson(path.join(runDir, 'config.json'), config);
  writeJson(path.join(runDir, 'metrics.json'), metrics);
  writeJson(path.join(runDir, 'transport.json'), transmitted.stats);
  writeJson(path.join(runDir, 'result.json'), {
    run_id: runId,
    output_path: outputPath,
    output_sha256: outputHash,
    content_match: contentMatch,
    validation,
  });
  fs.writeFileSync(path.join(runDir, 'encoded_payload.bin'), encoded.payload);

  return { runId, runDir, kind, metrics, outputPath };
}

async function runAll({ samplesDir, runsDir, transport, task = {} }) {
  ensureDir(runsDir);
  const inputs = {
    text: path.join(samplesDir, 'sample.txt'),
    image: path.join(samplesDir, 'sample.ppm'),
    video: path.join(samplesDir, 'sample.tvid'),
  };
  for (const inputPath of Object.values(inputs)) {
    if (!fs.existsSync(inputPath)) {
      throw new Error(`Missing sample: ${inputPath}`);
    }
  }
  const results = [];
  for (const [kind, inputPath] of Object.entries(inputs)) {
    results.push(await runOne({ kind, inputPath, runsDir, transport, task }));
  }
  return results;
}

module.exports = {
  runOne,
  runAll,
  createRunId,
  sha256,
};
